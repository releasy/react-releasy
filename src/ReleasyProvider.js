// @flow
import * as React from 'react';
import * as PropTypes from 'prop-types';
import { Environment } from 'relay-runtime';
import invariant from 'invariant';

import { getEnvironment } from './store';

type Props = {
  children: React.Node,
};

class ReleasyProvider extends React.PureComponent<Props> {
  static propTypes = {
    children: PropTypes.element.isRequired,
  };

  static childContextTypes = {
    environment: PropTypes.instanceOf(Environment).isRequired,
  };

  constructor(props: Props) {
    invariant(props.children, 'ReleasyProvider: missing "children"');

    super(props);
  }

  getChildContext() {
    return {
      environment: getEnvironment(),
    };
  }

  render() {
    return this.props.children;
  }
}

export default ReleasyProvider;
