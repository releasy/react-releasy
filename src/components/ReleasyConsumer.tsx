import * as React from 'react';

import GlobalContext, { GlobalContextValueType } from '../contexts/GlobalContext';

type ReleasyConsumerPropsType = {
  children: React.ReactNode,
};

const ReleasyConsumer = ({
  children,
}: ReleasyConsumerPropsType): React.ReactElement<GlobalContextValueType> => (
  <GlobalContext.Consumer>
    {(globalContextValue: GlobalContextValueType) =>
      React.Children.map(children, (child: React.ReactElement<any>) =>
        React.cloneElement(child, globalContextValue),
      )
    }
  </GlobalContext.Consumer>
);

export default ReleasyConsumer;
