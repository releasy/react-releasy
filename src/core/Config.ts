import { CacheConfig, RequestNode, UploadableMap, Variables, GraphQLResponse } from 'relay-runtime';
import invariant from 'invariant';

export interface LinkInterface {
  fetch: (
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables: UploadableMap,
  ) => Promise<GraphQLResponse>;
}

export interface CacheInterface {
  get: (queryID: string, variables: Variables) => GraphQLResponse;
  set: (queryID: string, variables: Variables, payload: GraphQLResponse) => void;
  clear: () => void;
  size: () => number;
}

type ConfigPropsType = {
  link: LinkInterface,
  cache?: CacheInterface,
  devTools?: boolean,
  networkLogger?: boolean,
};

class Config {
  link: LinkInterface;
  cache: CacheInterface;
  devTools: boolean;
  networkLogger: boolean;

  constructor(props: ConfigPropsType) {
    invariant(props, 'Config: missing props');
    invariant(props.link, 'Config: missing "link"');

    this.link = props.link;
    this.cache = props.cache || null;
    this.devTools = props.devTools || false;
    this.networkLogger = props.networkLogger || false;
  }
}

export default Config;
