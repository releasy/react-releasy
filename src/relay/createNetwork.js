// @flow
import { Network } from 'relay-runtime';
import RelayNetworkLogger from 'relay-runtime/lib/RelayNetworkLogger';

import type { CacheConfig, RequestNode, UploadableMap, Variables } from 'relay-runtime';

import { isMutation, mustForceFetch } from './utils';

import type { Config } from '../index';

const createNetwork = (config: Config): Network => {
  const fetchFn = config.networkLogger
    ? RelayNetworkLogger.wrapFetch(config.link.fetch)
    : config.link.fetch;

  return Network.create(
    async (
      request: RequestNode,
      variables: Variables,
      cacheConfig: CacheConfig,
      uploadables?: UploadableMap,
    ) => {
      const queryID = request.text;

      // no cache configured, just fetch the data from server
      if (!config.cache) {
        return fetchFn(request, variables, cacheConfig, uploadables);
      }

      // on mutations, clear cache and then fetch the data
      if (isMutation(request)) {
        config.cache.clear();
        return fetchFn(request, variables, cacheConfig, uploadables);
      }

      // verify if data exists on cache
      const fromCache = config.cache.get(queryID, variables);
      if (fromCache !== null && !mustForceFetch(cacheConfig)) {
        return fromCache;
      }

      const fromServer = await fetchFn(request, variables, cacheConfig, uploadables);
      if (fromServer) {
        config.cache.set(queryID, variables, fromServer);
      }

      return fromServer;
    }
  );
};

export default createNetwork;
