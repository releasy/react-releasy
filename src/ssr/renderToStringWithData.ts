
import { ReactElement } from 'react';
import * as ReactDOM from 'react-dom/server';

import getDataFromTree from './getDataFromTree';

const renderToStringWithData = async (component: ReactElement<any>): Promise<string> => {
  await getDataFromTree(component);
  return ReactDOM.renderToString(component);
};

export default renderToStringWithData;
