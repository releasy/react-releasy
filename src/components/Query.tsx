import * as React from 'react';
import { QueryRenderer } from 'react-relay';
import { GraphQLTaggedNode, Variables } from 'relay-runtime';

import { DataFrom } from '../types/react-relay.d';
import GlobalContext, { GlobalContextValueType } from '../contexts/GlobalContext';
import Config from '../core/Config';

export type QueryChildrenPropsType = {
  error?: Error,
  retry: Function,
  isFetching: boolean,
};

type QueryPropsType = {
  query: GraphQLTaggedNode,
  variables?: Variables | Function,
  children: (props: QueryChildrenPropsType) => JSX.Element,
  dataFrom?: DataFrom,
};

const Query = (props: QueryPropsType): React.ReactElement<any> => {
  const {
    query,
    variables,
    children,
    dataFrom,
  } = props;

  const getVariables = (): Variables => {
    if (typeof variables === 'function') {
      return variables(props);
    }

    return variables;
  };

  const getDataFrom = (config: Config): DataFrom => {
    if (!dataFrom && config.ssrMode === true) {
      return DataFrom.STORE_THEN_NETWORK;
    }

    return dataFrom;
  };

  return (
    <GlobalContext.Consumer>
      {({ environment, config }: GlobalContextValueType) =>
        <QueryRenderer
          environment={environment}
          query={query}
          variables={getVariables()}
          dataFrom={getDataFrom(config)}
          render={({ props: queryProps, error, retry }) => {
            return children({
              ...queryProps,
              error,
              retry,
              isFetching: !queryProps && !error,
            });
          }}
        />
      }
    </GlobalContext.Consumer>
  );
};

export default Query;
