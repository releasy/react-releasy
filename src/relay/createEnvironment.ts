import { installRelayDevTools } from 'relay-devtools';
import { Environment, RecordSource, Store } from 'relay-runtime';

import Config from '../core/Config';
import createNetwork from './createNetwork';

const createEnvironment = (config: Config): Environment => {
  if (config.devTools) {
    installRelayDevTools();
  }

  const network = createNetwork(config);

  const source = new RecordSource();
  const store = new Store(source);

  return new Environment({
    network,
    store,
  });
};

export default createEnvironment;
