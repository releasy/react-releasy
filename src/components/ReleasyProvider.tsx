import * as React from 'react';
import invariant from 'invariant';

import GlobalContext, {
  GlobalContextValueType,
  setConfig,
  setEnvironment,
} from '../contexts/GlobalContext';
import Config from '../core/Config';
import createEnvironment from '../relay/createEnvironment';

type ReleasyProviderPropsType = {
  config: Config,
  children: React.ReactNode,
};

type ReleasyProviderStateType = {
  globalContextValue: GlobalContextValueType,
};

class ReleasyProvider extends React.PureComponent<
  ReleasyProviderPropsType,
  ReleasyProviderStateType
> {
  constructor(props: ReleasyProviderPropsType) {
    super(props);

    invariant(props.config, 'ReleasyProvider: missing "config"');

    const { config } = props;
    const environment = createEnvironment(config);

    setConfig(config);
    setEnvironment(environment);

    this.state = {
      globalContextValue: {
        config,
        environment,
      },
    };
  }

  render() {
    const { children } = this.props;
    const { globalContextValue } = this.state;

    return (
      <GlobalContext.Provider value={globalContextValue}>
        {children}
      </GlobalContext.Provider>
    );
  }
}

export default ReleasyProvider;
