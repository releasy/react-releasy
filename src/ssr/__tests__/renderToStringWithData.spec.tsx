import * as React from 'react';

import { MeQuery } from '../../../tests/components';
import { createConfig } from '../../../tests/fixtures/core';
import { createQueryResponse } from '../../../tests/fixtures/relay';

import ReleasyProvider from '../../components/ReleasyProvider';
import renderToStringWithData from '../renderToStringWithData';

it('should render to string with data', async () => {
  global.fetch.mockResponse(
    JSON.stringify(createQueryResponse()),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  const config = createConfig({ ssrMode: true });

  const result = await renderToStringWithData(
    <ReleasyProvider config={config}>
      <MeQuery />
    </ReleasyProvider>,
  );

  expect(result).toMatchSnapshot();
});
