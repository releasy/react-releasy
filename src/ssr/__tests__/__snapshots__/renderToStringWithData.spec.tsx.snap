// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render to string with data 1`] = `"<div>fakeViewerId<!-- -->: <!-- -->John Doe</div>"`;
